### ISSUE TITLE
MathJax not loaded in plotly offline

### ISSUE THREAD

**Issue by rreusser** [plotly/plotly.py#515]
[`plotly.offline.plot`](https://github.com/plotly/plotly.py/blob/master/plotly/offline/offline.py#L314) fails (shows an unstyled plot after bailing out early due to errors) if latex is used since MathJax is not loaded in the default html. Short of getting fancy, it seems like `plotly.offline.plot` might need a manual option to inject MathJax from a cdn into the head tag.

**Comment by yankev** [comment]
Thanks Ricky, I'll check this out.

**Comment by physinet** [comment]
#105 addresses this as well. When using `plotly.offline.iplot`, the plot looks all right, except that LaTeX is not rendered in any of the text labels.

**Comment by mmlanger** [comment]
This is actually the only reason that stops me using plotly for my publication graphics.
No LaTeX in offline plotly is a serious issue, is there any news on this?
#444 and #445also adresses latex problems with plotly.

**Comment by szsiddiq** [comment]
FYI, sometimes after executing iplot, the LaTeX I have else where in my notebook (outside of plotly) breaks and stops rendering. I'm having trouble replicating this on sample data that I can share, but will update with code if I'm able to do so.

**Comment by jjaraalm** [comment]
This is not really an issue with plotly.py per-se, but is the result of issues upstream at plotly.js.

plotly.js does not encapsulate mathjax but instead relies on it being loaded globally and configured in a specific way. The problem is that there are multiple different output renderers for mathjax. Juypter globally loads mathjax with HTML output, but plotly.js requires SVG output. A quick and easy fix is just to change the output renderer for mathjax globally as this does not affect Juypter's markdown rendering. 

In order to address #445, mathjax should also be configured to set displayAlign = center since plotly.js explicitly sets this to left align. As far as I can tell in simple testing, the display alignment actually has no bearing on the output rendered by plotly.

Right now the code block I'm using to fix all these issues is:

```python

# Startup plotly
init_notebook_mode(connected=True)

# The polling here is to ensure that plotly.js has already been loaded before
# setting display alignment in order to avoid a race condition.
display(HTML(
    '<script>'
        'var waitForPlotly = setInterval( function() {'
            'if( typeof(window.Plotly) !== "undefined" ){'
                'MathJax.Hub.Config({ SVG: { font: "STIX-Web" }, displayAlign: "center" });'
                'MathJax.Hub.Queue(["setRenderer", MathJax.Hub, "SVG"]);'
                'clearInterval(waitForPlotly);'
            '}}, 250 );'
    '</script>'
))
```

**Comment by flo-compbio** [comment]
@jjaraalm's [piece of code](https://github.com/plotly/plotly.py/issues/515#issuecomment-263097892) (above) fixed the problem for me. Thank you!

**Comment by chriddyp** [comment]
Thanks for posting @jjaraalm ! I tried to add that snippet to the offline code but was getting some errors from Mathjax. Here is the PR: https://github.com/plotly/plotly.py/pull/683. 

Has anyone else been successful in adding this?

**Comment by yihengli** [comment]
Hi, are we still working on a fix for this issue?

I tried the code block posted by @jjaraalm but unfortunately it doesn't work for me. 
```python
import plotly
plotly.offline.init_notebook_mode(connected=True)
from IPython.core.display import display, HTML
# The polling here is to ensure that plotly.js has already been loaded before
# setting display alignment in order to avoid a race condition.
display(HTML(
    '<script>'
        'var waitForPlotly = setInterval( function() {'
            'if( typeof(window.Plotly) !== "undefined" ){'
                'MathJax.Hub.Config({ SVG: { font: "STIX-Web" }, displayAlign: "center" });'
                'MathJax.Hub.Queue(["setRenderer", MathJax.Hub, "SVG"]);'
                'clearInterval(waitForPlotly);'
            '}}, 250 );'
    '</script>'
))
```

I used nbconvert to html, but the latex will still show as $$ code $$. I am not sure if I used this code correctly. 

<img width="1114" alt="screen shot 2017-02-20 at 7 00 30 pm" src="https://cloud.githubusercontent.com/assets/13440386/23145769/e4592b04-f79e-11e6-990f-fe29627fa675.png">

**Comment by yihengli** [comment]
Another update: 
I found this code works well on Firefox, but not the same case on Chrome. 
<img width="1659" alt="screen shot 2017-02-20 at 7 13 29 pm" src="https://cloud.githubusercontent.com/assets/13440386/23145985/d73efe42-f7a0-11e6-89a0-e999b562ad3d.png">

However, in Chrome, if I use the inspect mode and then refresh the page, this code works again.
<img width="1339" alt="screen shot 2017-02-20 at 7 13 45 pm" src="https://cloud.githubusercontent.com/assets/13440386/23146003/ec506fbe-f7a0-11e6-8093-21bbaf623ae3.png">

**Comment by ianhbell** [comment]
Here is a MWE of latex working in offline mode.  A bit hacky, but it works.  Basically, add our own custom header with mathjax included, and then store our figure in a div.

```python
import plotly.offline as offline
import plotly.graph_objs as go
import os

trace1 = go.Scatter(
    x=[1, 2, 3, 4],
    y=[1, 4, 9, 16],
    name='$\\alpha_{1c} = 352 \\pm 11 \\text{ km s}^{-1}$'
)
data = [trace1]
layout = go.Layout(
    xaxis=dict(
        title='$\\sqrt{(n_\\text{c}(t|{T_\\text{early}}))}$'
    ),
    yaxis=dict(
        title='$d, r \\text{ (solar radius)}$'
    )
)
fig = go.Figure(data=data, layout=layout)

def with_jax(fig, filename):

    plot_div = offline.plot(fig, output_type = 'div')

    template = """
    <head>
    <script type="text/javascript" async
      src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_SVG">
    </script>
    </head>
    <body>
    {plot_div:s}
    </body>""".format(plot_div = plot_div)
    with open(filename, 'w') as fp:
        fp.write(template)
    os.startfile(filename)
with_jax(fig, 'cube.html')
```

**Comment by ianhbell** [comment]
Better would be a PR enabling the MathJAX CDN, which I am working on.

**Comment by mrclore** [comment]
I'm also interest in a solution to this problem. I do not use Jupyter notebooks, so the solution of jjaraalm does not work for me. The solution from ianhbell works fine for basic use, but it seems that it fails for parallel coordinates plot. Does anybody have a workaround for using latex in the labels of the parallel coordinate plots?

**Comment by ianhbell** [comment]
My MWE is no longer working, darn.

**Comment by SafwanAhmad** [comment]
It seems like a plotly.js or matjax bug. I was able to run Mathjax in the rest of my Jupyter notebook by changing
`MathJax.Hub.Config({messageStyle:"none",skipStartupTypeset:!0` to `MathJax.Hub.Config({messageStyle:"none",skipStartupTypeset:!1` in Plotly.js  in my HTML.

**Comment by connmay** [comment]
Unfortunately using the snippet from @jjaraalm now doesn't render any title when I try to use plot ly offline mode... Any ideas how to fix this? This occurs even after deleting the display(HTML(....)) code snippets and restarting the kernel. Before it would at least provide a title when I tried to use LATEX but now it only displays a title if I do not use any $ ... $ text in the title string. 

Any help would be appreciated!

**Comment by JacoGasp** [comment]
Same problem here.
Titles are not displayed if using $...$ math mode

**Comment by lucainnocenti** [comment]
I've been having the same (or a very similar one?) issue. Latex works in the jupyter notebook, but not always in the exported html version. Chrome still shows the equations fine (even though sometimes it doesn't and a reload of the page is required), while firefox doesn't. 

I found that a quick (if not very handy) fix is to add the following in the source of the exported html file, just after the MathJax configuration lines:

`<script type="text/javascript">MathJax.Hub.Queue(["Typeset", MathJax.Hub]);</script>`

I see the above behaviour already with the following MWE (plus any markdown cell with some latex in it):

```
import numpy as np
import plotly
import plotly.graph_objs as go
plotly.offline.init_notebook_mode(connected=False)
zs = np.random.randn(50, 50)
data = [go.Surface(z=zs)]
layout = go.Layout(
    autosize=False,
    width=500,
    height=500,
    margin=dict(
        l=0, r=0, b=0, t=0
    )
)
fig = go.Figure(data=data, layout=layout)
py.iplot(fig)
```

**Comment by sursu** [comment]
Any progress? Is anyone working on this issue?

**Comment by jonmmease** [comment]
Hi @sursu ,

I haven't started digging into this yet, but I definitely agree with everyone that it's very important.

One bit of good news is that Mathjax/latex support is already working in the [new offline static image export feature](https://github.com/plotly/plotly.py/pull/1120) that is coming in version 3.2.  So the publication use-case should finally be covered!

**Comment by jonmmease** [comment]
Hi All,

I've started digging into this a bit and I'm pretty confident we'll have latex/MathJax support working for `plotly.offline.plot` (standalone html/div) for version 3.3.0.  I'm also hopeful that we'll be able to add support for `plotly.offline.iplot` and `plotly.graph_objs.FigureWidget` at the same time, but these are a bit more complicated because Jupyter also relies on MathJax, and it requires MathJax be configured in a different global mode.

In case it's helpful for building workarounds in the meantime, here is the PR where I started experimenting with this: https://github.com/plotly/plotly.py/pull/1169.  This PR won't be merged in it's current form, but the basic concepts seem pretty sound.

Also in the meantime, `plotly.plotly.plot`, `plotly.plotly.iplot`, and the orca static image export already support latex/MathJax.

**Comment by jonmmease** [comment]
Done in #1243 and will be released in version 3.4.0.  LaTeX support requires no API changes for `iplot` or `FigureWidget` support.  An additional `include_mathjax` argument has been added to `plotly.offline.plot` to specify how MathJax should be included in the resulting html file/div.
